tests/test_core/test_core.py

```diff
@@ -0,0 +1,499 @@
+import anndata as ad
+import numpy as np
+import pandas as pd
+import pytest
+
+from ehrdata import EHRData
+from ehrdata.core.constants import R_LAYER_KEY
+
+
+def _shape_test(edata: EHRData, shape: tuple[int, int, int]):
+    assert edata.shape == shape
+
+    assert isinstance(edata.obs, pd.DataFrame)
+    assert len(edata.obs) == shape[0]
+    assert edata.n_obs == shape[0]
+
+    assert isinstance(edata.var, pd.DataFrame)
+    assert len(edata.var) == shape[1]
+    assert edata.n_vars == shape[1]
+
+    assert isinstance(edata.t, pd.DataFrame)
+    assert len(edata.t) == shape[2]
+    assert edata.n_t == shape[2]
+
+
+def _assert_fields_are_view(edata: EHRData):
+    assert edata.is_view
+    assert isinstance(edata.X, ad._core.views.ArrayView)
+    assert isinstance(edata.r, ad._core.views.ArrayView)
+    assert isinstance(edata.t, ad._core.views.DataFrameView)
+
+
+#################################################################
+### Test combinations of X, r, t during initialization
+#################################################################
+def test_ehrdata_init_vanilla_empty():
+    edata = EHRData()
+
+    _shape_test(edata, (0, 0, 0))
+
+    assert edata.X is None
+    assert edata.r is None
+
+    assert hasattr(edata, "obs")
+    assert edata.obs.shape == (0, 0)
+
+    assert hasattr(edata, "var")
+    assert edata.var.shape == (0, 0)
+
+    assert hasattr(edata, "t")
+    assert edata.t.shape == (0, 0)
```

Comment:
you can make all these tests much shorter by removing the redundant `hasattr` checks:

```suggestion
    edata = EHRData()

    _assert_shape_matches(edata, (0, 0, 0))
    assert edata.X is None
    assert edata.r is None
    assert edata.obs.shape == (0, 0)
    assert edata.var.shape == (0, 0)
    assert edata.t.shape == (0, 0)
```

Reply:
the comment applies to more tests, this one is just an example

Reply:
yep, I'm working on the others right now